### PR TITLE
DBZ-341 Ignore malformed messages coming from database history

### PR DESCRIPTION
**Pull request by jpechane**
This is just first step - to log history message before an exception is thrown to enable easier diagnostics

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 1]
```diff
@@ -198,14 +198,33 @@ protected void recoverRecords(Tables schema, DdlParser ddlParser, Consumer<Histo
...
                         Long lastOffset = offsetsByPartition.get(partition);
                         if (lastOffset == null || lastOffset.longValue() < record.offset()) {
-                            HistoryRecord recordObj = new HistoryRecord(reader.read(record.value()));
-                            records.accept(recordObj);
-                            logger.trace("Recovered database history: {}" + recordObj);
+                            if (record.value() == null) {
+                                logger.warn("Skipping null database history record. " +
+                                        "This is often not an issue, but if it happens repeatedly please check the '{}' topic.");
+                            }
+                            else {
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Minor nit: the `else` should be on the same line as the `}`.

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 2]
```diff
@@ -198,14 +198,33 @@ protected void recoverRecords(Tables schema, DdlParser ddlParser, Consumer<Histo
...
+                                HistoryRecord recordObj = new HistoryRecord(reader.read(record.value()));
+                                logger.trace("Recovering database history: {}" + recordObj);
+                                if (recordObj == null || !recordObj.isValid()) {
+                                    logger.warn("Skipping invalid database history record '{}'. " +
+                                            "This is often not an issue, but if it happens repeatedly please check the '{}' topic.",
+                                            recordObj,
+                                            topicName);
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Minor nit: put `topicName` on the same line as `recordObj`.

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 3]
```diff
@@ -198,14 +198,33 @@ protected void recoverRecords(Tables schema, DdlParser ddlParser, Consumer<Histo
...
+                                }
+                                else {
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Formatting of the `else` line, as mentioned earlier.

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 4]
```diff
@@ -198,14 +198,33 @@ protected void recoverRecords(Tables schema, DdlParser ddlParser, Consumer<Histo
...
+                                }
+                            }
                             offsetsByPartition.put(partition, new Long(record.offset()));
                             ++numRecordsProcessed;
                         }
                     } catch (IOException e) {
-                        logger.error("Error while deserializing history record", e);
+                        logger.error("Error while deserializing history record '{}'", record, e);
+                    }
+                    catch (final Exception e) {
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Again, formatting doesn't match existing code.

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 5]
```diff
@@ -198,14 +198,29 @@ protected void recoverRecords(Tables schema, DdlParser ddlParser, Consumer<Histo
...
+                            } else {
+                                HistoryRecord recordObj = new HistoryRecord(reader.read(record.value()));
+                                logger.trace("Recovering database history: {}" + recordObj);
+                                if (recordObj == null || !recordObj.isValid()) {
+                                    logger.warn("Skipping invalid database history record '{}'. " +
+                                            "This is often not an issue, but if it happens repeatedly please check the '{}' topic.",
+                                            recordObj, topicName);
+                                } else {
+                                    records.accept(recordObj);
+                                    logger.trace("Recovered database history: {}" + recordObj);
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: The log calls should pass the parameters as arguments, not using String concatenation. Looks like it was wrong in the original code.

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 6]
```diff
@@ -198,14 +198,29 @@ protected void recoverRecords(Tables schema, DdlParser ddlParser, Consumer<Histo
...
                         if (lastOffset == null || lastOffset.longValue() < record.offset()) {
-                            HistoryRecord recordObj = new HistoryRecord(reader.read(record.value()));
-                            records.accept(recordObj);
-                            logger.trace("Recovered database history: {}" + recordObj);
+                            if (record.value() == null) {
+                                logger.warn("Skipping null database history record. " +
+                                        "This is often not an issue, but if it happens repeatedly please check the '{}' topic.");
+                            } else {
+                                HistoryRecord recordObj = new HistoryRecord(reader.read(record.value()));
+                                logger.trace("Recovering database history: {}" + recordObj);
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: The log calls should pass the parameters as arguments, not using String concatenation. Looks like it was wrong in the original code.

File: debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java [hunk 7]
```diff
@@ -198,14 +198,29 @@ protected void recoverRecords(Tables schema, DdlParser ddlParser, Consumer<Histo
                         Long partition = new Long(record.partition());
                         Long lastOffset = offsetsByPartition.get(partition);
                         if (lastOffset == null || lastOffset.longValue() < record.offset()) {
-                            HistoryRecord recordObj = new HistoryRecord(reader.read(record.value()));
-                            records.accept(recordObj);
-                            logger.trace("Recovered database history: {}" + recordObj);
+                            if (record.value() == null) {
+                                logger.warn("Skipping null database history record. " +
+                                        "This is often not an issue, but if it happens repeatedly please check the '{}' topic.");
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java]: Missing topic param

File: debezium-core/src/test/java/io/debezium/relational/history/KafkaDatabaseHistoryTest.java [hunk 8]
```diff
@@ -175,4 +181,58 @@ protected void setLogPosition(int index) {
...
+            producer.send(noPositionRecord).get();
+            producer.send(invalidJSONRecord1).get();
+            producer.send(invalidJSONRecord2).get();
+        }
+
+        DdlParser recoveryParser = new DdlParserSql2003();
+
+        Tables recoveredTables = new Tables();
+        setLogPosition(1000);
+        history.recover(source, position, recoveredTables, recoveryParser);
```
Comment (gunnarmorling) [inline debezium-core/src/test/java/io/debezium/relational/history/KafkaDatabaseHistoryTest.java]: Shouldn't there be some kind of assertion? E.g. making sure a valid message can be consumed after running into some invalid ones? Or am I missing something here?

**Comment by gunnarmorling** [comment]
Thanks for adding the test, @jpechane. Squashed everything into one commit, rebased and applied.

**Pull request closed without merging**